README.md

```diff
@@ -4,142 +4,232 @@ Containers to be used for general purpose Data Science.
+  machine if the container itself is executed as root.
 
-- Docker provides the `USER` directive to change the active user (can create a user using the `adduser` command in linux, see code snippet above).
+- Docker provides the `USER` directive to change the active user (can create a
+  user using the `adduser` command in linux, see code snippet above).
 
-**Recommendation:** always set the user to a non-root user at the end of your Dockerfile by default using the `USER` directive in your Dockerfile.
+**Recommendation:** always set the user to a non-root user at the end of your
+Dockerfile by default using the `USER` directive in your Dockerfile.
 
 ### Do a multi-stage build
 
-- There are a couple of key reasons to building an image in this way.
+There are a couple of key reasons to building an image in this way.
 
-1. Reduced disk footprint - you can use one image to build one or more artifacts, then simply copy the relevant build artifact(s) out of the first image into a second image.
+- Reduced disk footprint - you can use one image to build one or more artifacts,
+  then simply copy the relevant build artifact(s) out of the first image into a
+  second image.
 
-2. Improved Security - You can use a complete image to build your application, then copy your application into a [distroless](https://github.com/GoogleContainerTools/distroless) image.
-   1. A distroless image is a very stripped down image that contains only your applications and its runtime dependencies (e.g. a Python image might contain a python interpreter in a virtual environment)
-   2. This implies that there are no shells, package managers, or other programs you would expect to find on a standard Linux distribution).
-   3. Improves security because there is much less attack surface area (i.e. there are fewer ways an attacker can perform a malicious act on a distroless image).
+- Improved Security - You can use a complete image to build your application,
+  then copy your application into a [distroless][] image.
+  - A distroless image is a very stripped down image that contains only your
+    applications and its runtime dependencies (e.g. a Python image might contain
+    a python interpreter in a virtual environment)
+  - This implies that there are no shells, package managers, or other programs
+    you would expect to find on a standard Linux distribution).
+  - Improves security because there is much less attack surface area (i.e. there
+    are fewer ways an attacker can perform a malicious act on a distroless
+    image).
 
-**Recommendation:** consider using a multistage build to reduce disk footprint, and, if applicable, consider copying your applications/build artifacts to a distroless image to improve security.
+**Recommendation:** consider using a multistage build to reduce disk footprint,
+and, if applicable, consider copying your applications/build artifacts to a
+distroless image to improve security.
 
```

Comment:
```suggestion

### Avoid cache-busting your Dockerfiles

- Docker tries to avoid building anything it knows has not changed, but it cannot always detect this properly.  When in situations where Docker always rebuilds layers unnecessarily, consider breaking your build into multiple images.  This way you can put your static content in the first image, then base your final image on that first image (`FROM myfirstimage:v123`).  This way you prevent Docker from trying to decide if your static content needs to be rebuilt (but note, if you do this and your static content does need rebuidling, you need to do that yourself!)

- Sometimes with python packages Docker will misunderstand whether a `requirements.txt` has changed, forcing a full  reinstall of your python packages even when they are cached.  One way around this is to add `RUN pip install mySlowPackages==0.1.2` explicitly near the top of the Dockerfile **in addition to in your requirements.txt file later in the build**.  The benefit of this is that those explicit calls to pip are easier for Docker to cache, so they won't be reinstalled as frequently.  And even though you do `pip install -r requirements.txt` later, pip itself will see that your package is already installed and skip it, meaning you save almost as much time as if Docker got the cache correct.  
```